provider/test-programs/sns-topic-replacement/main.go

```diff
@@ -0,0 +1,37 @@
+package main
+
+import (
+	"github.com/pulumi/pulumi-aws/sdk/v6/go/aws"
+	"github.com/pulumi/pulumi-aws/sdk/v6/go/aws/sns"
+	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
+	"github.com/pulumi/pulumi/sdk/v3/go/pulumi/config"
+)
+
+func main() {
+	pulumi.Run(func(ctx *pulumi.Context) error {
+		conf := config.New(ctx, "")
+		providerName := conf.Require("providerName")
+
+		// Create an AWS provider
+		provider, err := aws.NewProvider(ctx, providerName, &aws.ProviderArgs{})
+		if err != nil {
+			return err
+		}
+
+		topicName := conf.Require("topicName")
+
+		// Create an SNS Topic using the above created AWS provider
+		_, err = sns.NewTopic(
+			ctx,
+			topicName,
+			&sns.TopicArgs{Name: pulumi.StringPtrFromPtr(&topicName)},
+			pulumi.Provider(provider),
+			// pulumi.DeleteBeforeReplace(true),
```

Comment:
I'd expect this to not be needed if the resource has `DeleteBeforeReplace` set in `resources.go`.

The test however does fail when this is commented out:
https://github.com/pulumi/pulumi-aws/actions/runs/7557789851/job/20579141864?pr=3268

Is this a bug in the bridge?

Reply:
My expectation of the ResourceInfo `DeleteBeforeReplace` is that it essentially applies the equivalent of the `DeleteBeforeReplace` resource option every time the resource is called.

This does not seem to be the case, since commenting out the live above passes the test.

Is my expectation wrong here or is this a bug?

Reply:
I might be misunderstanding something here, since it looks like `Diff` is never called for the SNS topic resource.

I verified this by adding a panic in https://github.com/pulumi/pulumi-terraform-bridge/blob/6e4059f3759494d0cef96baec289ba2092740dd2/pkg/tfbridge/provider.go#L793

Reply:
My comment reminded me of this https://github.com/pulumi/pulumi-aws/issues/1923#issuecomment-1858544873, I wonder if something similar is happening.

Reply:
Yeah this option is advertised as such but scanning the implementation code I'm not convinced it's working actually so if you have evidence here that it's not working then indeed very much seems like we need to open and fix a bug in the bridge.

```
	DeleteBeforeReplace bool        // if true, Pulumi will delete before creating new replacement resources.
```

Reply:
Diff should definitely be called for pulumi preview/pulumi up on sns Topic, I suspect something is odd with your test setup?

Reply:
That's sounds very plausible 👍